File: gnu/gcc/libgomp/testsuite/libgomp.fortran/appendix-a/a.39.1.f90
! { dg-do run }

      SUBROUTINE SKIP(ID)
      END SUBROUTINE SKIP
      SUBROUTINE WORK(ID)
      END SUBROUTINE WORK
      PROGRAM A39
        INCLUDE "omp_lib.h"      ! or USE OMP_LIB
        INTEGER(OMP_LOCK_KIND) LCK
        INTEGER ID
        CALL OMP_INIT_LOCK(LCK)
!$OMP PARALLEL SHARED(LCK) PRIVATE(ID)
          ID = OMP_GET_THREAD_NUM()
          CALL OMP_SET_LOCK(LCK)
          PRINT *, "My thread id is ", ID
          CALL OMP_UNSET_LOCK(LCK)
          DO WHILE (.NOT. OMP_TEST_LOCK(LCK))
            CALL SKIP(ID)     ! We do not yet have the lock
                              ! so we must do something else
          END DO
          CALL WORK(ID)       ! We now have the lock
                              ! and can do the work
          CALL OMP_UNSET_LOCK( LCK )
!$OMP END PARALLEL
        CALL OMP_DESTROY_LOCK( LCK )
        END PROGRAM A39

File: gnu/gcc/libgomp/testsuite/libgomp.c/appendix-a/a.39.1.c
/* { dg-do run } */

#include <stdio.h>
#include <omp.h>
void
skip (int i)
{
}

void
work (int i)
{
}
int
main ()
{
  omp_lock_t lck;
  int id;
  omp_init_lock (&lck);
#pragma omp parallel shared(lck) private(id)
  {
    id = omp_get_thread_num ();
    omp_set_lock (&lck);
    /* only one thread at a time can execute this printf */
    printf ("My thread id is %d.\n", id);
    omp_unset_lock (&lck);
    while (!omp_test_lock (&lck))
      {
	skip (id);		/* we do not yet have the lock,
				   so we must do something else */
      }
    work (id);			/* we now have the lock
				   and can do the work */
    omp_unset_lock (&lck);
  }
  omp_destroy_lock (&lck);
  return 0;
}
